### DISCUSSION TITLE
Compatible with WASM target?

### DISCUSSION THREAD

**Discussion by coder0xff** [webrtc-rs/webrtc#394] [Q&A]
Hi. I read that UDP sockets aren't in WASM. Will this work as expected when targeting WASM with Rust? For additional context, I'm writing a Flutter application that uses a Rust library, which I'm also writing, that will implement a bespoke internet protocol. Thanks!

**Comment by k0nserv** [comment]
It's not possible to use this with WASM as it exists in the context of the web. For WASM runtimes that support WASI or some other way to perform socket IO it might be possible.

If you can implement UDP socket I/O you can provide a muxed connection via `SettingEngine::set_udp_network`.

**Reply by walterjj** [reply]
Does that mean it should work in a node.js server context?

**Reply by teohhanhui** [reply]
The `webrtc` / `webrtc-*` crates currently use `tokio`'s `"net"` feature, which does not work on the `wasm32-wasip2` target:

https://github.com/rust-lang/rust/issues/130323

And even when https://github.com/tokio-rs/tokio/pull/6893 is merged, it still won't work on stable Rust:

https://github.com/rust-lang/rust/issues/130323#issuecomment-2350379144

**Comment by InDieTasten** [comment]
While not an answer to your actual question, maybe some other folks would like to know an alternative solution to this problem:

When talking about Browsers as WASM target, they already ship with webrtc. You can implement your browser client application using rust-wasm and then interop to the JS webrtc API for webrtc scenarios. Admittedly not a perfect solution, but it's what we got in browsers at the moment.
